radix-common/src/data/manifest/model/manifest_address_reservation.rs

```diff
@@ -14,6 +10,109 @@ use crate::*;
 #[must_use]
 pub struct ManifestAddressReservation(pub u32);
 
+//========
+// resolution
+//========
+
+/// This is for use with the [`ResolvableManifestAddressReservation`] trait.
+/// Implementers should panic if a proof cannot be found.
+pub trait NamedManifestAddressReservationResolver {
+    fn assert_address_reservation_exists(&self, address_reservation: ManifestAddressReservation);
+    fn resolve_address_reservation(&self, name: &str) -> ManifestAddressReservation;
+}
+
+/// This trait is intended to be used as an `impl` argument in helper methods
+/// operating on manifests, to resolve a [`ManifestAddressReservation`] from a name, an id,
+/// or from itself.
+///
+/// The resolution process relies on a [`NamedManifestAddressReservationResolver`] which can
+/// provide a lookup to/from names.
+pub trait ResolvableManifestAddressReservation {
```

Comment:
Why do we have a `Resolvable` trait for each resolvable type that we have instead of having a single `ResolvableInto` trait? 

I tested this out a little bit and I think that I very much like having a single trait. Here's a code example that compiles:

```rust
use radix_common::prelude::*;
use radix_transactions::prelude::*;

/// An example resolver trait.
pub trait Resolver {
    fn resolve_address_reservation(
        &self,
        name: &str,
    ) -> ManifestAddressReservation;

    fn resolve_named_address(&self, name: &str) -> DynamicGlobalAddress;

    fn resolve_bucket(&self, name: &str) -> ManifestBucket;

    fn resolve_proof(&self, name: &str) -> ManifestProof;
}

// Traits to resolve from and into a type. They're written this way so that they
// mirror the `From` and `Into` traits in the standard library, we even provide
// an `Into` implementation for all `From` implementations in a very similar way
// to the standard library.

pub trait TryResolveFrom<T>
where
    Self: Sized,
{
    type Error;

    fn try_resolve_from(
        value: T,
        resolver: &impl Resolver,
    ) -> Result<Self, Self::Error>;
}

pub trait TryResolveInto<T> {
    type Error;

    fn try_resolve_into(
        self,
        resolver: &impl Resolver,
    ) -> Result<T, Self::Error>;
}

/* Generic impl of Into for all types that impl From */
impl<T, U> TryResolveInto<U> for T
where
    U: TryResolveFrom<T>,
{
    type Error = U::Error;

    fn try_resolve_into(self, resolver: &impl Resolver) -> Result<U, Self::Error> {
        U::try_resolve_from(self, resolver)
    }
}

/* All types that implement ResolveFrom also implement an infallible TryResolveFrom */
impl<T, U> TryResolveFrom<T> for U where U: ResolveFrom<T> {
    type Error = std::convert::Infallible;

    fn try_resolve_from(value: T, resolver: &impl Resolver) -> Result<Self, Self::Error> {
        Ok(value.resolve_into(resolver))
    }
}

pub trait ResolveFrom<T> {
    fn resolve_from(value: T, resolver: &impl Resolver) -> Self;
}

pub trait ResolveInto<T> {
    fn resolve_into(self, resolver: &impl Resolver) -> T;
}

/* Generic impl of Into for all types that impl From */
impl<T, U> ResolveInto<U> for T
where
    U: ResolveFrom<T>,
{
    fn resolve_into(self, resolver: &impl Resolver) -> U {
        U::resolve_from(self, resolver)
    }
}

/* Generic impl of ResolvableFrom for all types that impl From */
impl<T, U> ResolveFrom<T> for U where U: From<T> {
    fn resolve_from(value: T, _: &impl Resolver) -> Self {
        value.into()
    }
}

// An example implementation of the trait and how we can use it in resolutions.
impl<T> ResolveFrom<T> for ManifestBucket
where
    T: AsRef<str>,
{
    fn resolve_from(value: T, resolver: &impl Resolver) -> Self {
        resolver.resolve_bucket(value.as_ref())
    }
}

// We get to implement Option resolution for all `Option` types by default and
// without the need for us to implement it for each type.
impl<T, U> ResolveFrom<Option<T>> for Option<U>
where
    U: ResolveFrom<T>,
{
    fn resolve_from(value: Option<T>, resolver: &impl Resolver) -> Self {
        value.map(|value| value.resolve_into(resolver))
    }
}

// Testing out the implementation of the traits to make sure that they do
// compile.
fn example_fn() {
    fn func_that_takes_resolvable_bucket(_: impl ResolveInto<ManifestBucket>) {}
    fn func_that_takes_an_optional_resolvable_bucket(
        _: impl ResolveInto<Option<ManifestBucket>>,
    ) {
    }

    func_that_takes_resolvable_bucket("hello world!");
    func_that_takes_an_optional_resolvable_bucket(Some("hello world!"));
}
```

What I like about this is:
1. We have a single resolvable trait which seems very elegant.
2. We have an abstraction that's very similar to that of the standard library in the form of `ResolvableFrom` and `ResolvableInto` and even the default `Into` impl for all `From` that we see in the standard library.
3. It allows us to write generic implementations for things like `Option<T>`, `Result<O, E>` and other types (see the implementation for options in the example I provided above).
4. I very very much like how similar this is to the standard library and how the amount of generic implementations we can provide for all of these traits. The interface of the manifest builder would primarily make use of the try traits. As an example `fn consume_bucket(bucket: impl TryResolveInto<ManifestBucket>);`

Reply:
I like the idea in principle, but I think I'd want it to be more generic, i.e. we have two traits:
* `Resolve<X>` - which doesn't need a resolver.
* `ResolveFrom<T, X>` - which has the resolve method take a `Resolver<T, X>`.

I imagine we _might_ hit some issues though with overlapping blanket implementations and things; but we could try it.

Reply:
What are your thoughts on that ^? 👍

Reply:
Have just tried it, and I get conflicts between these:
```rust
impl<T, K, X, E> ResolveFrom<K, X> for T
where
    T: TryInto<X, Error = E>,
    E: Debug,
{
    fn resolve(self, _resolver: &impl KeyedResolver<K, X>) -> X {
        self
            .try_into()
            .unwrap_or_else(|err| panic!("Value was not a valid {}: {err:?}", type_name::<X>()))
    }
}
```
and:
```rust
impl<'a, X> ResolveFrom<&'a str, X> for &'a str {
    fn resolve(self, resolver: &impl KeyedResolver<&'a str, X>) -> X {
        resolver.resolve_from_key(self)
    }
}

impl<'a, X> ResolveFrom<&'a str, X> for &'a String {
    fn resolve(self, resolver: &impl KeyedResolver<&'a str, X>) -> X {
        resolver.resolve_from_key(self.as_str())
    }
}

impl<'a, X> ResolveFrom<&'a str, X> for String {
    fn resolve(self, resolver: &impl KeyedResolver<&'a str, X>) -> X {
        resolver.resolve_from_key(self.as_str())
    }
}
```

I think we might need to implement the blanket alongside each `X`:
```rust
impl<T, K, X, E> ResolveFrom<K, X> for T
where
    T: TryInto<X, Error = E>,
    E: Debug,
{
    fn resolve(self, _resolver: &impl KeyedResolver<K, X>) -> X {
        self
            .try_into()
            .unwrap_or_else(|err| panic!("Value was not a valid {}: {err:?}", type_name::<X>()))
    }
}
```

Reply:
Ahh even worse, this blanket implementation can't work -- because `KeyedResolve` is "foreign", and therefore a third crate could implement `KeyedResolve` and `TryInto<X>` on its type `T`, breaking coherence.

This explains why I originally put the trait alongside its definitions I guess?
```rust
/// ```rust
/// # struct X;
/// # type K = str;
/// # use radix_rust::prelude::*;
/// impl<T, E> KeyedResolve<X, K> for T
/// where
///     T: TryInto<X, Error = E>,
///     E: Debug,
/// {
///     fn resolve(self, _resolver: &impl KeyedResolver<K, X>) -> X {
///         self
///         .try_into()
///         .unwrap_or_else(|err| panic!("Value was not a valid {}: {err:?}", core::any::type_name::<X>()))
///     }
/// }
/// ```
```

Reply:
But perhaps we can just ignore the blanket implementation from `TryInto<X, Error = E>` in this case, and just implement the definite ones we need? This might work